FHEM/00_SIGNALduino.pm

```diff
@@ -2921,7 +2923,40 @@ sub SIGNALduino_Attr(@) {
 		return "Note, please use other userReadings names.\nReserved names from $name are: cc1101_config, cc1101_config_ext, cc1101_patable"
 			if ($aVal =~ /cc1101_(?:config(?:_ext)?|patable)(?:\s|{)/);
 	}
- 	return ;
+	elsif ($aName eq "rfmode")	# change receive mode
+	{
+		my $oldAttrib = AttrVal($name, 'rfmode', 'SlowRF');
+
+		if ( ($aVal ne $oldAttrib) && ($hash->{DevState} eq 'initialized') && (InternalVal($hash->{NAME},"cc1101_available",0) == 1) ) {
+			$hash->{logMethod}->($name, 3, "$name: Attr, $aName set to $aVal (please check protocollist)");
+
+			if ($aVal ne 'SlowRF') {
+				for my $id (@{$hash->{mnIdList}}) {
+					my $rfmode=$hash->{protocolObject}->checkProperty($id,'rfmode',-1);
+
+					if ($rfmode eq $aVal) {
+						$hash->{logMethod}->($name, 5, qq[$name: Attr, rfmode found on ID=$id]);
+						my $register=$hash->{protocolObject}->checkProperty($id,'register', -1);
+
+						if ($register != -1) {
+							$hash->{logMethod}->($name, 5, qq[$name: Attr, register settings exist on ID=$id ]);
+
+							for my $i (0...scalar(@{$register})-1) {
+								$hash->{logMethod}->($name, 5, "$name: Attr, $aName - write value " . @{$register}[$i]);
+								my $argcmd = sprintf("W%02X%s",hex(substr(@{$register}[$i],0,2)) + 2,substr(@{$register}[$i],2,2));
+								main::SIGNALduino_AddSendQueue($hash,$argcmd);
+							}
+							main::SIGNALduino_WriteInit($hash);  # Todo: from AOK/OOK -> xFSK need Reset | xFSK -> AOK/OOK no Reset ?? Strobe Commands ??
+							last;                                # found $rfmode, exit loop
+						}
+					}
+				}
+			} else {
+				main::SIGNALduino_AddSendQueue($hash,'e');
```

Comment:
Ist das eine wirklich gute Idee, einen factory Reset auf dem uC  zu machen? Sind wir uns da ganz sicher, dass wir alle Register neu setzen?
Was mache ich, wenn ich Register selbst angepasst habe, aber den Betriebsmodi mal umstellen muss?

Reply:
>Was mache ich, wenn ich Register selbst angepasst habe, aber den Betriebsmodi mal umstellen muss?

Dann benenne mir bitte den Fall, was du genau als "normaler Benutzer" von Hand anpassen würdest.

Dein Argument ist berechtigt ABER das haben wir generell nicht vorgesehen. Wieso wollen wir dann gerade hier dies berücksichten? 

Wenn ich die Bandbreite oder patable ändere im normalen Modus ASK/OOK und mache ein factory Reset, so wird diese auch zurück gestellt. Wo fangen wir an und wo hören wir auf?

Es geht hier in erster Linie darum, das der Benutzer wieder zu "normal" zurück kommt.

Ich würde vorschlagen, man kann darüber nachdenken ein zusätzliches Attribut zu erweitern wo ein Benutzer sein selbst eingerichtetes Register speichert. Alle anderen Überlegungen führen vermutlich dazu, das wir den Einbau ins unermesslicher "durchkauen".

Reply:
Ich hatte über den Fall nachgedacht, dass jemand vielleicht Somfy 433 Mhz empfangen möchte.
Im Forum kann ich mich Erinneren, dass die Frequenz ein wenig angepasst werden musste, weil Somfy nicht auf 433,92 sendet. Also z.B. hat der Benutzer auf 433.8 Mhz gestellt und auch an der Bandbreite etwas gedreht.

Auf 433 Mhz läuft jetzt alles, also z.B. Funksteckdosen und Somfy.

Empfang von 868 Mhz / FSK ist hier nicht drin, aber könnte es sein, dass er vielleicht eine PCA Steckdose hat die er damit schalten möchte? Also stellt er dazu vermutlich das Attribut um und anschließend wieder zurück ? Mit einem doif oder sowas in der Art.

Im Microcontroller merken wir uns diverse Register, wenn wir diese nur temporär umstellen. Wäre hier nicht so einfach möglich. :(

Reply:
Deine Bedenken sind richtig, nur sehe ich diese etwas weit hergegriffen.

Das ein User, den Empfangsmodus umstellt mit einem DoIF .... hm, also da gehört mehr als Basiswissen dazu. Sobald ein Benutzer auf 433 und 868 aktiv empfangen möchte, so ist immer der Weg zu 2 Empfängern ratsam.

Wenn es dir rein um die Einstellung geht, das wir nicht einfach auf `raw e` zurückfallen, so könnte man sich die Einstellungen merken in einem Attribut. Sobald der User solche benuterspezifischen Anpassungen hat, so kann in einem Attribut dies hinterlegt werden und beim zurückschreiben genutzt werden.

*leichsinnig würde ich sagen, lass die Kirche im Dorf*

Reply:
Ich gehe die Sachen halt vorher gerne mal durch.

Zumindest weiss ich dann, dass wir ein potentielles Problem einbauen.
Ich würde die Diskussion beenden.

Dass wir die Register aus FHEM heraus schreiben bekommen wir vermutlich nicht so einfach hin, da wir ja nur in mnIDList suchen und für den "default" die Register nicht hinterlegt haben.

Reply:
Kann es sein, das du kein Interesse hast daran xFSK zu nutzen und einzubinden? Ich sehe das nicht als gravierendes Problem.

Default gibt es im uC. Ein default separat definieren müssen wir nicht, es sei denn der Benutzer hat eine sehr abweichende Konfiguration. (Speicherung als Attribut)

Das wir nur MNList durchsuchen hat den Hintergrund, das alles andere unnötig wäre. Ich denke fast, Dir fehlt dort ein wenig Input vielleicht. Du wolltest, das wir die Register dort ablegen wo die Protokolle definiert sind weil letztendlich das ein Protokoll oder eine Hardware wiederspiegelt.

Der Ansatz aus FHEM die Register zu setzen haben auch andere Empfänger. Vorteil von der Verankerung im Modul / Protokolldatei der Register, das wir die Firmware nicht aufblähen.

Eine potenzielle Gefährdung sehe ich nicht.

Reply:
Im Moment wüsste ich nicht ob mir persönlich xFSK was bringt.

Ich will das nicht blockieren, nur verstehen.
Die Firmware aufblähen möchte ich auch nicht. Daher auch die kurze Überlegung, ob die Register auch aus dem Modul gesetzt werden könnten. Das wird aber auf Anhieb nichts.

Idee für später wäre also, ein Attribut zu haben in dem die Default Register stehen. Immer wenn das definiert ist, machen wir dann kein factory reset auf dem uC sondern setzen die Register, welche im Modul definiert sind.
Klingt für mich prinzipiell verständlich.

Vielleicht kann man die Register ja auch direkt in rFMode ablegen.

Reply:
Ich habe eine Varinate wo wir das Ganze umsetzen könnten wie folgt:

- zusätzliches Attribut cc1101_reg_user, wo selbst definierte Register mit Werten gespeichert werden
- zusätzliches Set Befehl cc1101_reg_user, wo die Werte aus dem Attribut geschrieben werden

Somit kann man sich ein komplettes oder einen Teil des Register speichern.

Sobald man nun den Modulationstyp umstellt bzw. zurück zu ASK/OOK, so wird automatisch geschaut ob das Attribut vorhanden ist, wenn ja nutzt man diese Werte, wenn nein, dann raw e.

Vorteile, man könnte hinter dem Setbefehl welcher derzeit ohne Argumente ist, die Modulationen ASK_OOK bzw. FSK als Auswahl schaffen, was aber wiederum einen anderen Attributnamen zur Folge hätte.

Das Thema ist Basis für den gesamt PR für FSK, so suchte ich nun nach einem Weg, welcher für die Zukunft sicher scheint und auch praktikabel. Der Vorteil liegt so auf der Hand, das man mit raw e zu den FW defauls kommt und mit dem Attribut, kann man sich selbst sogar dies definieren.

Was hällst du von der Variante?

Ich möchte ungern dir ewig lange PRś bieten wo du dann nicht durchsiehst ;-) Deswegen hier Fortsetzung im Dialog ;-)

Reply:
Finde ich vom Ansatz her gut.  

In dem Attribut kann man sich Register hinterlegen und der Set Befehl schreibt dann nur die Register aus dem Attribut nehme ich an.

Aus dem Namen des Attributes geht nicht hervor, dass es die Default Register sind welche beim Setzen von slowRF verwendet werden. Fällt dir da eine Bezeichnung ein?

Bezüglich des set Befehles, würde dieser ja die dort hinterlegten Register schreiben. Das geht aus dem Namen noch nicht hervor. Vielleicht etwas in dieser Form:
`cc1101_reg_default`

Wäre aber spannend, was ist der Default wenn ich nicht slowRF als RF Mode verwende?

Wenn Du dieses Feature in diesem PR behandeln möchtest, wird es leider noch ein paar Tage dauern bis ich verstanden habe, welche Sitation wie funktionieren soll.

Ich denk schon dauern daran, wie ich dafür tests schreiben kann :)

Reply:
Ich werde mal versuchen dir etwas Licht ins Dunkle zu bringen.

>Aus dem Namen des Attributes geht nicht hervor, dass es die Default Register sind welche beim Setzen von slowRF verwendet werden. Fällt dir da eine Bezeichnung ein?

Das Attribut nennt sich derzeit `cc1101_reg_user`. Den Bezug auf default bitte nicht vornehmen. Es handelt sich hierbei um ein Attribut wo beliebige Anpassungen, welche als Default definiert werden vom Benutzer hinterlegt werden können. Hierbei ist die Modulationsart auch uninteressant.

>Bezüglich des set Befehles, würde dieser ja die dort hinterlegten Register schreiben. Das geht aus dem Namen noch nicht hervor. Vielleicht etwas in dieser Form:
cc1101_reg_default

Hier gilt das gleiche. Das selbst hinterlegte Register `cc1101_reg_user` aus dem Attribut muss nicht zwangläufig default sein. Es können auch Werte zum Testen sein, welche ich nach einem `raw e` zurück schicken möchte.

Wir haben im uC einen Default festgelegt. Dieser kann jederzeit mit `raw e` hergestellt werden. Dafür benötigen wir kein Attribut wo wir dies zusätzlich speichern müssen.
1) Wenn der User nun aufgrund der Gegebenheiten sein Register anpasst, so speichert er diesen Wert in dem Attribut.
2) Das Attribut ist unabhängig von der Modulation
3) Die Vorgabewerte um spezielle Hardware zu empfangen und ggf zu verarbeiten kommt in die ProtocolDatei unter den Punkt Register.

Wenn der Benutzer den RFmode ändert von Standard (ASK/OOK) auf Bsp. Lacrosse oder Kopp, so nutzen wir die Registerwerte aus der Protokolldatei. Sollte da ebenso bei dem einen die Frequenz etwas variieren, so kann nach der Einstellung auf den Bedarf das Attribut erneut damit gefüllt werden.

Wir geben sozusagen das "Grundgerüst" für den Benutzer vor und alles andere kann über das zusätzliche Attribut verfeinert werden. Das Umschalten der verschiedenen Moden, welche wir definierten in der Protokolldatei ist somit unabhängig.

>Wenn Du dieses Feature in diesem PR behandeln möchtest, wird es leider noch ein paar Tage dauern bis ich verstanden habe, welche Sitation wie funktionieren soll.

Ich denke, du hast es bestimmt nun besser verstanden :-)
Das was auf jedenfall als gegeben ist, das wir bei unterschiedlichen Modulationen immer den CC1101 "individuell" einstellen weil man da die Vorgaben macht für Datarate oder SyncDaten unsw. Der Positive Effekt bei diesem Feature, welches wir für xFSK benötigen um Testregister schnell umzuschalten, so profitieren die Benutzer welche wie Bsp. bei SOMFY abweichende ASK/OOK Werte nutzen im Register. Sie müssen nach dem Einmaligen setzen im Attribut nach einem Raw e nur noch EINMAL den Setbefehl setzen und fertig.

Die Beschreibung für das Attribut & den Setbefehl erfolgt dann in der Commandref, so das der User somit klar kommen sollte. (Wer mit Registern arbeitet beim CC1101, so gehe ich davon aus, das dort auch mehr Wissen vorhanden ist)

Ich überlege "nur" ob die Benennung `cc1101_reg_user` gut ist oder ob nicht vielleicht `cc1101_register` passender wäre.

PS: in dem Attribut kann man beliebig viele Registerwerte speichern. Wenn ich ständig nur ein Registerwert anpasse abweichend von raw e, so speicherte ich nur eines. Sollte ich 5 immer ändern wollen, so speicherte ich 5.

PS2: Das jetzige Vorgehen hält uns immer noch die Option offen, in ferner ZUKUNFT .... dies vielleicht mal anzupassen von
`cc1101_reg_user:norArg` zu Bsp: `cc1101_reg_user:Typ1,Typ2` ....

Nun sollte der Hintergrund nachvollziehbar sein.